src/qibo/core/terms.py

```diff
@@ -0,0 +1,273 @@
+import sympy
+from qibo import gates, K
+from qibo.config import raise_error, EINSUM_CHARS
+
+
+class HamiltonianTerm:
+    """Single term in a :class:`qibo.core.hamiltonians.SymbolicHamiltonian`.
+
+    This Hamiltonian is represented by a sum of such terms which are held in
+    its ``.terms`` attribute which is a list.
+
+    Args:
+        matrix (np.ndarray): Full matrix corresponding to the term representation
+            in the computational basis. Has size (2^n, 2^n) where n is the
+            number of target qubits of this term.
+        q (int): Target qubit ids.
+    """
+
+    def __init__(self, matrix, *q):
+        self.target_qubits = tuple(q)
+        self._gate = None
+        self.hamiltonian = None
+        if not (matrix is None or isinstance(matrix, K.qnp.numeric_types) or
+                isinstance(matrix, K.qnp.tensor_types)):
+            raise_error(TypeError, "Invalid type {} of symbol matrix."
+                                   "".format(type(matrix)))
+        self._matrix = matrix
+
+    @property
+    def matrix(self):
+        return self._matrix
+
+    @property
+    def gate(self):
+        """Qibo gate that implements the action of the term on states."""
+        if self._gate is None:
+            self._gate = gates.Unitary(self.matrix, *self.target_qubits)
+        return self._gate
+
+    def exp(self, dt):
+        """Matrix exponentiation of the term."""
+        return K.qnp.expm(-1j * dt * self.matrix)
+
+    def expgate(self, dt):
+        """Unitary gate implementing the matrix exponentiation of the term."""
+        return gates.Unitary(self.exp(dt), *self.target_qubits)
+
+    def merge(self, term):
+        """Creates a new term by merging the given term to the current one.
+
+        The target qubits of the given term should be a subset of the target
+        qubits of the current term.
+        """
+        matrix = K.np.kron(term.matrix, K.eye(2 ** (len(self) - len(term))))
+        matrix = K.np.reshape(matrix, 2 * len(self) * (2,))
```

Comment:
I must admit that that K.np, K.qnp or simply K can be misleading. Do you still need this triple distinction?
Ideally with cupy/numpy we could use directly K, I believe.

Reply:
Btw, we are doing `self.qnp = self.construct_backend("numpy")` so most likely we could replace that with `K.np`.

Reply:
> I must admit that that K.np, K.qnp or simply K can be misleading. Do you still need this triple distinction?

I think the distinction between K and K.np is still useful because in some cases we prefer to use numpy instead of the backend. This was first introduced for the tensorflow backend because we noticed that numpy is faster than tensorflow in performing small matrix operations. In such cases we use K.np to force use of numpy regardless of the activated backend. These operations are for example gate multiplications of small (2x2 or 4x4 matrices) in gate fusion, or the exponentiation of small matrices for Trotterized evolution which is used in the current PR. I have not tested but it is likely that numpy is still faster than numba/cupy for such small operations so performance-wise having K.np should be helpful.

Regarding K.np vs K.qnp, the difference is that the first is the actual numpy while the second is qibo's `NumpyBackend`. There are small API differences between the two, but I am not sure if these are enough for this distinction. I can have a look over the whole qibo code and perhaps we could remove one of the two. For these specific lines you note both K.qnp and K.np would work the same.

Reply:
Ok, fair enough.